questionpy_server/collector/local_collector.py

```diff
@@ -201,32 +121,114 @@ def __init__(self, directory: Path, indexer: 'Indexer'):
+
+            :param directory: The directory.
+            :return: A generator of paths.
+            """
+
+            for file in Path(directory).glob('*.qpy'):
+                if file.is_file():
+                    yield file
+
+        async def add_package(pkg_hash: str, pkg_path: Path) -> None:
+            """
+            Adds a package to the map and registers it in the indexer.
+
+            :param pkg_hash: The hash of the package.
+            :param pkg_path: The path of the package.
+            """
+
+            self.map.insert(pkg_hash, pkg_path)
+            await self.indexer.register_package(pkg_hash, pkg_path, self)
+
+        async def remove_package(pkg_path: Path) -> None:
+            """
+            Removes a package from the map and unregisters it from the indexer.
+
+            :param pkg_path: The path of the package.
+            """
+
+            if not (pkg_hash := self.map.pop(pkg_path)):
+                return
+            if not (packages := self.map.get(pkg_hash)) or len(packages) == 0:
+                # There are no other packages with the same hash - unregister it.
+                await self.indexer.unregister_package(pkg_hash, self)
+
+        if not self._lock:
+            self._lock = Lock()
+
+        async with self._lock:
+            # If no snapshot exists, use EmptyDirectorySnapshot to get all files as created.
+            old_snapshot = self._snapshot or EmptyDirectorySnapshot()
+            new_snapshot = await to_thread(DirectorySnapshot, str(self.directory), recursive=False,
+                                           listdir=directory_iterator)
+            difference = DirectorySnapshotDiff(old_snapshot, new_snapshot)
+
+            for path in difference.files_created:
+                package_path = Path(path)
+                package_hash = await to_thread(calculate_hash, package_path)
+                await add_package(package_hash, package_path)
+
+            for path in difference.files_deleted:
+                package_path = Path(path)
+                await remove_package(package_path)
+
+            for path in difference.files_modified:
+                package_path = Path(path)
+                package_hash = await to_thread(calculate_hash, package_path)
+                await remove_package(package_path)
+                await add_package(package_hash, package_path)
+
+                self._log.warning("Package %s was modified. This will cause unexpected behavior if the package is "
+                                  "currently read by a worker.", package_path)
+
+            # We need to remove every old path before inserting new ones to avoid conflicts when paths get swapped.
+            entries = []
+            for old_path, new_path in difference.files_moved:
+                # Remove old path and save the package hash.
+                if existing_hash := self.map.pop(Path(old_path)):
+                    entries.append((existing_hash, Path(new_path)))
```

Comment:
Warum rufst du hier nicht direkt `self.map.insert()` auf?

Reply:
Falls zwei Pakete, `p1.qpy` und `p2.qpy`, im Ordner existieren und zwischen zwei Snapshots die Dateinamen getauscht werden, `p1.qpy → p2.qpy` und `p2.qpy → p1.qpy`, würde es durch den sofortigen Aufruf von `self.map.insert()` zu einem Konflikt bei `self.map.pop()` kommen.
In `self.map` würde zuerst `p1.qpy` entfernt und dann der neue Dateiname `p2.qpy` hinzugefügt werden, was den alten Eintrag von `p2.qpy` überschreibt. Bei der nächsten Iteration würde dann `p2.qpy` entfernt werden, was allerdings jetzt das soeben hinzugefügte und nicht das alte Paket ist.
Aus diesem Grund müssen zuerst alle alten Dateinamen aus der `self.map` entfernt und danach die Neuen hinzugefügt werden.
Diesen Fall habe ich auch mit[ einem Test](https://github.com/questionpy-org/questionpy-server/blob/e52b8d724163284135eb2665f3ada89d843abfc4/tests/collector/test_local_collector.py#L267-L293) abgedeckt.

Reply:
Ah verstehe. Gut dass du auch an diesen Sonderfall denkst. :)